kernel/src/common/deferred_call_mux.rs

```diff
@@ -0,0 +1,207 @@
+//!
+//! The `deferred_call_mux_clients` array size determines how many
+//! [DeferredCallHandle](crate::common::deferred_call_mux::DeferredCallHandle)s
+//! may be registered with the Mux.
+//! When no more slots are available,
+//! `deferred_call_mux.register(some_client)` will return `None`.
+//!
+//! ```
+//! # use core::cell::Cell;
+//! # use kernel::common::cells::OptionalCell;
+//! # use kernel::static_init;
+//! use kernel::common::deferred_call_mux::{
+//!     DeferredCallMux,
+//!     DeferredCallMuxClient,
+//!     set_global_mux,
+//! };
+//!
+//! let deferred_call_mux_clients = unsafe { static_init!(
+//!     [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>); 1],
+//!     [(Cell::new(false), OptionalCell::empty())]
+//! ) };
+//! let deferred_call_mux = unsafe { static_init!(
+//!     DeferredCallMux,
+//!     DeferredCallMux::new(deferred_call_mux_clients)
+//! ) };
+//! assert!(unsafe { set_global_mux(deferred_call_mux) }, true);
+//!
+//! # struct SomeCapsule;
+//! # impl SomeCapsule {
+//! #     pub fn new(_mux: &'static DeferredCallMux) -> Self { SomeCapsule }
+//! #     pub fn set_deferred_call_handle(
+//! #         &self,
+//! #         _handle: kernel::common::deferred_call_mux::DeferredCallHandle,
+//! #     ) { }
+//! # }
+//! # impl DeferredCallMuxClient for SomeCapsule {
+//! #     fn call(
+//! #         &self,
+//! #         _handle: kernel::common::deferred_call_mux::DeferredCallHandle,
+//! #     ) { }
+//! # }
+//! #
+//! // Here you can register custom capsules, etc.
+//! // This could look like:
+//! let some_capsule = unsafe { static_init!(
+//!     SomeCapsule,
+//!     SomeCapsule::new(deferred_call_mux)
+//! ) };
+//! some_capsule.set_deferred_call_handle(
+//!     deferred_call_mux.register(some_capsule).expect("no deferred call slot available")
+//! );
+//! ```
+
+use crate::common::cells::OptionalCell;
+use crate::common::deferred_call::DeferredCall;
+use core::cell::Cell;
+
+pub const DEFERRED_CALL_MUX_TASK: usize = 31; // highest bit in usize reserved for deferred call mux
+
+static mut DEFERRED_CALL_MUX: Option<&'static DeferredCallMux> = None;
+
+/// Sets a global `DeferredCallMux` instance
+///
+/// This is required before any deferred calls can be retrieved.
+/// It may be called only once. Returns `true` if the mux was successfully
+/// registered.
+pub unsafe fn set_global_mux(mux: &'static DeferredCallMux) -> bool {
+    // If the returned reference is identical to the mux argument,
+    // it is set in the option. Otherwise, a different mux is
+    // already registered and may not be replaced.
+    (*DEFERRED_CALL_MUX.get_or_insert(mux)) as *const _ == mux as *const _
+}
+
+/// Call the globally registered mux
+///
+/// Returns true if a Mux was registered and has been called.
+/// This function needs to be called by the underlying deferred
+/// call implementation in the `chip` crate.
+pub unsafe fn call_global_mux() -> bool {
+    DEFERRED_CALL_MUX.map(|mux| mux.call()).is_some()
```

Comment:
Because Tock's design is limited to single threaded code execution, this unsafe part shouldn't cause too much trouble, right?

Reply:
I would claim that the most important thing as such executions can't be interleaved e.g, interrupts in `not-reentrant` functions such as this one, but because the kernel is event-driven (run-to-completion) it should be safe AFAIK.

However, `tock` more or less forces this design unless you manually implement `unsafe Send + Sync`

/cc @alevy please correct me if I'm wrong